pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+		{
+			self.total_stake += rewards;
+			if need_update_free_stake {
+				self.free_stake += rewards;
+			}
+
+			for vault_staker in &self.vault_stakers {
+				let mut vault = ensure_vault::<T>(*vault_staker)
+					.expect("vault in vault_stakers should always exist: qed.");
+				let nft_id = Pallet::<T>::merge_or_init_nft_for_staker_inner(
+					self.cid,
+					vault.user_id.clone(),
+				)
+				.expect("merge nft shoule always success: qed.");
+				let nft = Pallet::<T>::get_nft_attr_inner(self.cid, nft_id)
+					.expect("get nft attr should always success: qed.");
+				let vault_shares = nft.shares.to_fixed();
+				let stake_ratio = match vault_shares.checked_div(self.total_shares.to_fixed()) {
+					Some(ratio) => BalanceOf::<T>::from_fixed(&ratio),
+					None => continue,
+				};
+				let settled_stake = bmul(stake_ratio, &rewards.to_fixed());
+				vault.basepool.distribute_reward::<T>(settled_stake, false);
+			}
+		}
+	}
+
+	pub fn pallet_id<T>() -> T
+	where
+		T: Encode + Decode,
+	{
+		(b"basepool")
+			.using_encoded(|b| T::decode(&mut TrailingZeroInput::new(b)))
+			.expect("Decoding zero-padded account id should always succeed; qed")
+	}
+
+	impl<T: Config> Pallet<T>
+	where
+		T: mining::Config<Currency = <T as Config>::Currency>,
+		BalanceOf<T>: sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+		T: pallet_uniques::Config<CollectionId = CollectionId, ItemId = NftId>,
+	{
+		#[frame_support::transactional]
+		pub fn contribute<F>(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			userid: T::AccountId,
+			amount: BalanceOf<T>,
+			maybe_settle_method: Option<F>,
+		) -> Result<BalanceOf<T>, DispatchError>
+		where
+			F: FnMut(
+				&BasePool<T::AccountId, BalanceOf<T>>,
+				&mut NftAttr<BalanceOf<T>>,
+				T::AccountId,
+			),
+		{
+			ensure!(
+				// There's no share, meaning the pool is empty;
+				base_pool.total_shares == Zero::zero()
+                // or there's no trivial `total_stake`, meaning it's still operating normally
+                || base_pool.total_stake > Zero::zero(),
+				Error::<T>::PoolBankrupt
+			);
+			let nft_id = Self::merge_or_init_nft_for_staker(base_pool, userid.clone())?;
+			// The nft instance must be wrote to Nft storage at the end of the function
+			// this nft's property shouldn't be accessed or wrote again from storage before set_nft_attr
+			// is called. Or the property of the nft will be overwrote incorrectly.
+			let mut nft = Self::get_nft_attr(base_pool, nft_id)?;
+			// NFT should always settled befroe adding/ removing.
+
+			if let Some(mut settle_method) = maybe_settle_method {
+				settle_method(base_pool, &mut nft, userid.clone());
+			}
+
+			//maybe_settle_nft_slash is only for stakepool
+			//Self.maybe_settle_nft_slash(&mut nft, userid.clone());
+
+			let shares = Self::add_stake_to_new_nft(base_pool, userid.clone(), amount);
+			// Lock the funds
+
+			//ledger should be processed outside basepool (only for stakepool)
```

Comment:
If you intend to note to the caller, move the comment to the function docstring.

Reply:
not needed, just removed